File: examples/java/gmsub_disp.java
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/** 
 * @file gmsub_disp.java
 *
 * A simple example Java GMSEC subscriber 
 *
 */


import gov.nasa.gsfc.gmsec.api.*;
import gov.nasa.gsfc.gmsec.api.util.Log;

import java.util.ArrayList;


class PublishDispatcher extends Callback 
{
	public void onMessage(Connection conn, Message message) 
	{
		Log.info("[PublishDispatcher.onMessage] Received:\n" + message.toXML());

		Log.info("Press the enter key to exit");
	}
}


public class gmsub_disp implements Example 
{
	Config     config;
	Connection connection;
	ArrayList<SubscriptionInfo> info = new ArrayList<SubscriptionInfo>();

	gmsub_disp(Config config) throws ExampleException
	{
		this.config = config;

		Util.initialize(config);

		if (!Util.hasConnectionType(config))
		{
			printUsage();
			throw new ExampleException("Invalid option");
		}
	}


	private void printUsage()
	{
		System.err.println("\nusage: java gmsub_disp connectionType=<middleware> " +
		                   "[ <parameter>=<value> ]\n" +
		                   "\n\tNote that the parameter 'connectiontype' is required. "  +
		                   "\n\tThe rest of other parameters are optional.\n" +
		                   "\n\tserver=<server name> " +
		                   "(required if middleware is not bolt/MB locally)" +
		                   "\n\tsubject=<subject name>" +
		                   "\n\tsubject.<N>=<sub-subject name>" +
		                   "\n\t\tNote: N must be 1 or greater" +
		                   "\n\t\tFor example, subject.1=GMSEC.A subject.2=GMSEC.B" +
		                   "\n\tloglevel=<log level>" +
		                   "\n\tcfgfile=<config_filepath.xml>" +
		                   "\n\nFor more information, see API User's Guide\n\n");
	}


	public boolean run()
	{
		boolean result = true;

		// determine what subjects to listen to
		java.util.List<String> subjects = new java.util.ArrayList<String>(10);

		String value = Util.get(config,"SUBJECT");
		if (value != null) {
			subjects.add(value);
		}
		else {
			for (int i = 1; ; ++i) {
				String tmp = Util.get(config, "SUBJECT." + i);
				if (tmp != null)
					subjects.add(tmp);
				else
					break;
			}
		}

		if (subjects.isEmpty()) {
			String subject = Util.get(config, "SUBJECT", "GMSEC.>");
			subjects.add(subject);
		}
		else {
			subjects.add("GMSEC.TERMINATE");
		}

		// Output GMSEC API version
		Log.info(Connection.getAPIVersion());

		try
		{
			// Establish a connection
			connection = Util.openConnection(config);

			// Output middleware version
			Log.info("Middleware version = " + connection.getLibraryVersion());

			// Subscribe
			for (String subject : subjects)
			{
				Log.info("subscribing to '" + subject + "' with PublishDispatcher");
				info.add(connection.subscribe(subject, new PublishDispatcher()));
			}

			Log.info("Starting Auto-Dispatch");
			connection.startAutoDispatch();

			Log.info("Press the enter key to exit");

			try {
				while (System.in.available() == 0)
				{
					Util.millisleep(100);
				} 
			}
			catch (Exception e) {
			}

			Log.info("Stopping Auto-Dispatch");
			connection.stopAutoDispatch();
		}
		catch (GMSEC_Exception e)
		{
			Log.error("GMSEC_Exception: " + e.toString());
			result = false;
		}
		finally
		{
			try
			{
				cleanup();
			}
			catch (GMSEC_Exception e)
			{
				Log.error("GMSEC_Exception: " + e.toString());
				result = false;
			}
		}

		return result;
	}


	public boolean cleanup() throws GMSEC_Exception
	{
		if (connection != null)
		{
			for(int i = info.size()-1; i >= 0; i-- )
			{
				Log.info("Unsubscribing from " + info.get(i).getSubject());
				connection.unsubscribe(info.get(i));
				info.remove(i);
			}
			Util.closeConnection(connection);
		}

		return true;
	}


	public static void main(String args[]) 
	{
		try
		{
			Config config = new Config(args);

			Util.addToConfigFromFile(config);

			new gmsub_disp(config).run();
		}
		catch (ExampleException e)
		{
			System.err.println(e.toString());
		}
	}
}

File: examples/perl/gmsub_disp.pl
#
# Copyright 2007-2016 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.
#

# 
# gmsub_disp.pl
# 
# A perl example demonstrating the subscription abilities of 
# GMSEC.  This example involves the use of Callback functionality,
# in combination with the AutoDispatcher for receipt of messages.
# 


use strict;
use Util;

use libgmsec_perl;

*isa = \&UNIVERSAL::isa;

# Define the PublishCallback Callback Package
{
	package PublishCallback;
	use base 'libgmsec_perl::Callback';

	sub new
	{
		# Initialize the base Package (libgmsec_perl::Callback)
		my $class = shift;
		my $self = $class->SUPER::new();

		# Set the iteration limit and the shared Condition object to coordinate
		# with the main thread's (caller of startAutoDispatch) logic
		($self->{ITERATIONS}, $self->{CONDITION}) = @_;

		# Initialize the count to 0
		$self->{COUNT} = 0;

		return $self;
	}

	sub onMessage
	{
		my ($self, $conn, $message) = @_;

		libgmsec_perl::LogInfo("[PublishCallback::onMessage] Received:\n" . $message->toXML());

		my $iterations = $self->{ITERATIONS};
		my $done = ($iterations > 0 ? 1 : 0) && ++$self->{COUNT} >= $iterations;
		$done = $done || ($message->getSubject() eq "GMSEC.TERMINATE" ? 1 : 0);

		# If done, signal the main thread to exit its loop and shut down the auto-dispatcher
		if ($done)
		{
			$self->{CONDITION}->signal($libgmsec_perl::Condition::USER);
		}
	}
} # end PublishCallback definition


{
	my $task = bless({});
	Util::driver($task, @ARGV);
}


sub printUsage
{
	my ($self, $config) = @_;
	my $hasConnectionType = Util::hasConnectionType($config);
	return 0 if ($hasConnectionType);

	print "\nusage: perl gmsub_disp.pl connectiontype=<middleware> [ <parameter>=<value> ]" .

	"\n\n\tNote that the parameter 'connectiontype' is required." .
	"\n\tThe rest of other parameters are optional." .

	"\n\n\tserver=<server name> (required if middleware is not bolt/MB locally)" .
	"\n\tsubject=<subject name>" .
	"\n\tsubject.N=<sub-subject name>" .
	"\n\t\tNote: N must be 1 or greater" .
	"\n\t\tFor example, subject.1=GMSEC.A subject.2=GMSEC.B"  .
	"\n\titerations=<iterations>" .
	"\n\tloglevel=<log level>" .
	"\n\tcfgfile=<config_filename.xml>" .
	"\n\nFor more information, see API User's Guide\n";

	return 1;
}


sub run
{
	my ($self, $config) = @_;

	# Determine run-time settings
	my @subjects = Util::determineSubjects($config);
	my $iterations = Util::getInteger($config, "ITERATIONS", 0);

	if ($iterations > 0)
	{
		libgmsec_perl::LogInfo("Waiting for up to " . $iterations . " messages");
	}

	# Output GMSEC API version
	libgmsec_perl::LogInfo(libgmsec_perl::Connection::getAPIVersion());

	my $connection;

	eval
	{
		# Create the Connection
		$connection = libgmsec_perl::Connection::create($config);
		$self->{CONNECTION} = $connection;

		# Connect
		$connection->connect();

		# Output middleware version
		libgmsec_perl::LogInfo("Middleware version = " . $connection->getLibraryVersion());

		# Output example program-specific information
		libgmsec_perl::LogInfo("Using subjects: " . join(", ", @subjects));
		if ($iterations > 0)
		{
			libgmsec_perl::LogInfo("Receiving " . $iterations . " message(s)");
		}

		# Subscribe using a PublishCallback
		my $condition = libgmsec_perl::Condition->new();
		my $publishCallback = PublishCallback->new($iterations, $condition);
		$publishCallback->DISOWN();

		# Subscribe
		for my $subject (@subjects)
		{
			libgmsec_perl::LogInfo("Subscribing to " . $subject);
			push(@{$self->{SUBINFO}}, $connection->subscribe($subject, $publishCallback));
		}

		my $autoMutex = libgmsec_perl::AutoMutex->new($condition->getMutex());
		libgmsec_perl::LogInfo("Starting AutoDispatcher");
		if ($connection->startAutoDispatch())
		{
			my $conditionCode = -1; # Set the condition code to a value that the Condition enum could never be
			while ($conditionCode != $libgmsec_perl::Condition::USER)
			{
				$conditionCode = $condition->wait();
			}
		}
		else
		{
			libgmsec_perl::LogError("Failed to start AutoDispatcher");
		}
	};
	if (isa($@, 'libgmsec_perl::Exception'))
	{
		libgmsec_perl::LogError($@->what());
	}
	elsif($@)
	{
		libgmsec_perl::LogError($@);
	}

	$connection->stopAutoDispatch();

	$self->cleanup();
}


sub cleanup
{
	my ($self) = @_;

	my $connection = $self->{CONNECTION};

	if ($connection)
	{
		for my $subInfo (@{$self->{SUBINFO}})
		{
			eval
			{
				libgmsec_perl::LogInfo("Unsubscribing from " . $subInfo->getSubject());
				$connection->unsubscribe($subInfo);
			};
			if (isa($@, 'libgmsec_perl::Exception'))
			{
				libgmsec_perl::LogError($@->what());
			}
			elsif($@)
			{
				libgmsec_perl::LogError($@);
			}
		}

		libgmsec_perl::Connection::destroy($connection);

	}

	libgmsec_perl::Connection::shutdownAllMiddlewares();
}


1;

__END__
